ssz/src/main/java/tech/pegasys/teku/ssz/schema/SszUnionSchema.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.ssz.schema;
+
+import java.util.List;
+import java.util.stream.IntStream;
+import tech.pegasys.teku.ssz.SszData;
+import tech.pegasys.teku.ssz.SszUnion;
+import tech.pegasys.teku.ssz.schema.impl.SszUnionSchemaImpl;
+import tech.pegasys.teku.ssz.sos.SszLengthBounds;
+
+public interface SszUnionSchema<SszUnionT extends SszUnion> extends SszSchema<SszUnionT> {
+  int SELECTOR_SIZE_BYTES = 1;
+  int NONE_VALUE_SELECTOR = 0;
+
+  static SszUnionSchema<?> create(List<SszSchema<?>> childrenSchemas) {
+    return new SszUnionSchemaImpl(childrenSchemas);
+  }
+
+  List<SszSchema<?>> getChildrenSchemas();
+
+  /**
+   * Returns the child schema for selector.
+   *
+   * @throws IndexOutOfBoundsException if selector >= getTypesCount
+   */
+  default SszSchema<?> getChildSchema(int selector) {
+    return getChildrenSchemas().get(selector);
+  }
+
+  default int getTypesCount() {
+    return getChildrenSchemas().size();
+  }
+
+  SszUnionT createFromValue(int selector, SszData value);
+
+  @Override
+  default boolean isFixedSize() {
+    return false;
+  }
+
+  @Override
+  default int getSszFixedPartSize() {
+    return SELECTOR_SIZE_BYTES;
+  }
+
+  @Override
+  default SszLengthBounds getSszLengthBounds() {
+    return IntStream.range(0, getTypesCount())
+        .mapToObj(this::getChildSchema)
+        .map(SszType::getSszLengthBounds)
+        .reduce(SszLengthBounds::or)
+        .orElseThrow()
+        .addBytes(SELECTOR_SIZE_BYTES)
+        .ceilToBytes();
```

Comment:
nit: We found that we were doing a lot of GC because of the dynamically calculated `SszLengthBounds` objects since they get created on every gossip message.

Can we move this into the implementation and either calculate it once in the constructor or memoize it?

Reply:
Good point! Done: eb1fe24e